### PR TITLE
feat(qgrads): added parametrization and module quantization

### PR DESCRIPTION
**Pull request by LeoBuron**
In this feature the current state of the quantized grads got heavily changed. This is due to several reasons. One beeing that the data of the fixed point config was needed in function calls. Thus, the change of the position of the tensors was not possible after generating the function. Also, the support for other Optimizers was not possible and thus, the use of built in learning rate scheduler from pytorch.

This commit features
1. param_quantizations that register a parametrization for each parameter of the module. Also, a version for STE
2. module_quantizations those are used to quantize the output of a layer with a custom autograd function. They are available for fixedpoint with forw only and forw+back quantizations.
3. A quantized optimizer factory for updating parametrized parameters according to their quantization scheme

To Discuss:
1. Each Layer now has a math_ops field. This is the module used for quantizing the output of the layer and applying the autograd function. I do not have a better name for that field and it does relatively the same as our previous implementations. If you have a better name for that field, let me know. 
2. In my opinion this is a breaking change, but as I am the only one using my package, I am not sure, if it makes sense to make this a breaking change for versioning. Thus, I propose to not mark it as a breaking change 

For Review I recommend the following order:
1. layers in ```base_modules```
2. FixedPointConfigV2
3. _quantize_to_fixedpoint
4. Param Quantization
5. Autograd + Module Quantization

### CODE REVIEW

**Comment by LeoBuron** [comment]
@glencoe Ruff check throws an error, but the solution they propose is an absolute import instead of a relative one. However, that breaks the unit tests. What should I do? 
I propose correct unit-tests are more important than the failed check from ruff.

**Comment by glencoe** [comment]
> @glencoe Ruff check throws an error, but the solution they propose is an absolute import instead of a relative one. However, that breaks the unit tests. What should I do? I propose correct unit-tests are more important than the failed check from ruff.

I'll have a look

File: elasticai/creator/nn/quantized_grads/__init__.py [hunk 1]
```diff
@@ -1,12 +1,3 @@
...
 
-__all__ = [
-    "MathOperations",
-    "QuantizedSGD",
-    "QuantizeForw",
-    "QuantizeForwBackw",
-    "MathOperationsForw",
-    "MathOperationsForwBackw",
-]
+from .quantised_optim import factory_quantised_optimizer
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/__init__.py:3]: throughout the creator i think we use quantize instead of quantise

File: elasticai/creator/nn/quantized_grads/base_modules/batchnorm2d.py [hunk 2]
```diff
@@ -1,58 +1,22 @@
...
     To keep the weights quantized use only optimizers that apply a quantized update.
-    IMPORTANT: use only in-place operators for your quantize implementations
     """
 
     def __init__(
         self,
-        operations: MathOperations,
-        weight_quantization: Callable[[Tensor], None],
-        bias_quantization: Callable[[Tensor], None],
+        math_ops: Module,
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/base_modules/batchnorm2d.py:17]: also referring to your question, if it serves the same purpose as `MathOperations` and it just happens to additionally be a module, i'd just stick to calling it operations here.

File: elasticai/creator/nn/quantized_grads/base_modules/conv2d.py [hunk 3]
```diff
@@ -1,65 +1,21 @@
...
-            dtype=dtype,
-        )
-
-
-class MathOperations(Quantize, Protocol): ...
-
-
-class Conv2d(_Conv2d, QuantizedParameters):
+class Conv2d(TorchConv2d):
     """This module implements a 2d convolution.
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/base_modules/conv2d.py:10]: just a comment: you can skip `"This module"` and just state `"A 2d Convolution."` which tbh, we can tell by the name already. Can you provide a more concrete description here? What's different from torch.nn.Conv2d?

File: elasticai/creator/nn/quantized_grads/fixed_point/__init__.py [hunk 4]
```diff
@@ -1,13 +1,42 @@
...
-)
-from .quantize_to_fixed_point import (
-    quantize_to_fxp_stochastic_ as quantize_to_fxp_stochastic_,
-)
-from .two_complement_fixed_point_config import FixedPointConfigV2 as FixedPointConfigV2
+from .two_complement_fixed_point_config import FixedPointConfigV2
+
+from .param_quantization import (QuantizeParamSTEToFixedPointHTE as QuantizeTensorToFixedPointHTE,
+    QuantizeParamSTEToFixedPointStochastic as QuantizeTensorToFixedPointStochastic,
+ ) # adding this Modules again under a different name to make clear that it can be used for Tensor quantization.
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/__init__.py:22]: I suggest adding that to the short doc string instead and just export it once for this purpose.
If you can see that the different exports will change for different reasons, that'd be a better reason to export them like this. Though in that case explicitly subclassing should probably be preferred.

File: elasticai/creator/nn/quantized_grads/fixed_point/autograd.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
+from typing import Callable
+
+import torch
+from torch import Tensor
+
+from .quantize_to_fixed_point import quantize_to_fxp_hte, quantize_to_fxp_stochastic
+
+
+def factory_autograd_func(
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/autograd.py:9]: how about calling this `create_autograd_func` or `make_autograd_func`?

The term `factory_autograd_func` sounds misleading to me. I'd expect it to give me a factory for autograd functions.

Also i think the term factory is misused like crazy in most of python community, which leads to more and more people interpreting it differently. Heck, i've been guilty of this myself, mostly due to laziness.

This is what all but some python people will refer to as a factory

https://refactoring.guru/design-patterns/abstract-factory

File: elasticai/creator/nn/quantized_grads/base_modules/batchnorm2d.py (hunk 2 above)
Comment (glencoe) [reply elasticai/creator/nn/quantized_grads/base_modules/batchnorm2d.py:17]: A very brief comment explaining why this is a module here, will probably be useful in the future

File: elasticai/creator/nn/quantized_grads/fixed_point/quantize_to_fixed_point.py [hunk 6]
```diff
@@ -1,94 +1,75 @@
...
+from torch import Tensor
 
-from .two_complement_fixed_point_config import FixedPointConfigV2
 
-
-def _clamp_(number: torch.Tensor, fxp_conf: FixedPointConfigV2) -> None:
+class Round(torch.autograd.Function):
     """
-    Inplace clamp values to twos complement.
+    This function implements an STE for rounding.
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/quantize_to_fixed_point.py:7]: To provide another example for a short description, i suggest to reword this to something like
  
  `"Round deterministically to nearest neighbour with STE."`

File: elasticai/creator/nn/quantized_grads/fixed_point/quantize_to_fixed_point.py [hunk 7]
```diff
@@ -1,94 +1,75 @@
...
-
-
-def _round_to_fixed_point_hte_(
-    number: torch.Tensor, fxp_resolution: torch.Tensor
-) -> None:
-    """
-    Implements round half to even with torch rounding function. Inplace operation.
-    """
-    number.mul_(fxp_resolution).round_().div_(fxp_resolution)
+def round_tensor(x: Tensor) -> Tensor:
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/quantize_to_fixed_point.py:19]: top ;)

File: tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_batchnorm.py
Comment (glencoe) [inline tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_batchnorm.py:1]: I really like how this became easier to setup and read 👍

File: tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_linear.py [hunk 8]
```diff
@@ -1,34 +1,22 @@
...
-    FixedPointConfigV2,
-    quantize_to_fxp_stochastic,
-    quantize_to_fxp_stochastic_,
+    FixedPointConfigV2, QuantizeParamToFixedPointHTE, QuantizeParamToFixedPointStochastic,
+    QuantizeForwHTE
 )
 
 
-def test_linear_fxp_init():
+def test_conv1d_fxp_init():
```
Comment (glencoe) [inline tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_linear.py:10]: Isn't this test missing an assert?

File: tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py [hunk 9]
```diff
@@ -63,82 +61,25 @@ def test_round_determinstic_to_fixed_point():
...
-    _clamp_(x, conf)
-    expected = torch.Tensor([-2.0, -2.0, -1.0, 0.0, 1.0, 1.75, 1.75])
-    assert torch.equal(x, expected)
-
-
 def test_quantize_determinstic_to_fixed_point():
     conf = FixedPointConfigV2(total_bits=4, frac_bits=2)
     x = torch.arange(-3, 3.1, step=0.33, dtype=torch.float32)
-    actual = quantize_to_fxp_hte(x, conf)
+    actual = quantize_to_fxp_hte(number=x,
```
Comment (glencoe) [inline tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py:79]: I guess i missed that in this functions definition: Was there a specific reason to not just pass the FixedPointConfigV2 to `quantize_to_fxp_hte`?

File: tests/unit_tests/nn/quantized_grads/fxpmodel_loss_optim_test.py [hunk 10]
```diff
@@ -1,51 +1,45 @@
...
-from elasticai.creator.nn.quantized_grads.quantized_sgd import QuantizedSGD
-
+from elasticai.creator.nn.quantized_grads.fixed_point import FixedPointConfigV2, QuantizeParamToFixedPointHTE, \
+    QuantizeForwHTEBackwHTE
+from elasticai.creator.nn.quantized_grads.quantised_optim import factory_quantised_optimizer
 
 class Test:
     def test_pred(self, pred: torch.Tensor) -> None:
-        expected_pred = torch.Tensor([1.25, 3.75])
+        expected_pred = torch.Tensor([0.5, 2])
```
Comment (glencoe) [inline tests/unit_tests/nn/quantized_grads/fxpmodel_loss_optim_test.py:14]: What was the reason for this change?

I'm asking to ensure this shouldn't just be an additional test case instead of replacing one.

**Review by glencoe — CHANGES_REQUESTED** [review]
Got a few questions and suggestions, but nothing major.

The only things i'm actually requesting to change is 

* [ ] adhere to 100 char restriction for commit message body width
* [ ] sort imports with ruff to make CI linter happy

The rest looks good to me.

File: tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_linear.py (hunk 8 above)
Comment (LeoBuron) [reply tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_linear.py:10]: Kind of. I just wanted to test, if I can initalize a module. This doesn't test anything that goes beyond that, which is kinda bad. But I am too lazy to fix now

File: tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py (hunk 9 above)
Comment (LeoBuron) [reply tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py:79]: You don't want the tensors to stay in the fixed_point config, but in the module. Thus, you need to pass them one by one.

File: tests/unit_tests/nn/quantized_grads/fxpmodel_loss_optim_test.py (hunk 10 above)
Comment (LeoBuron) [reply tests/unit_tests/nn/quantized_grads/fxpmodel_loss_optim_test.py:14]: It couldn't test the behaviour of backward rounding, because they were already too big. Thus, I changed the values to test that too. But you are right. I should have added another test.

File: tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_linear.py (hunk 8 above)
Comment (glencoe) [reply tests/unit_tests/nn/quantized_grads/fixed_point/base_modules/test_linear.py:10]: You could make the name more explicit, that would suffice already from my perspective: e.g., `test_can_initialize_conv1d_fxp`

File: tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py (hunk 9 above)
Comment (glencoe) [reply tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py:79]: I don't see why that should be the case. Doing smth like this should have the same effect.

```python
def fn(self, thing_owning_tensors):
  self.my_tensor = thing_owning_tensors.my_tensor
```

**Comment by glencoe** [comment]
Please rebase instead of creating a merge commit, if possible. Having a mostly linear history will make our future life much easier.

File: tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py (hunk 9 above)
Comment (LeoBuron) [reply tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py:79]: Sure you could do that, but that would mean that the module where I can only register tensors as buffers would need to pack the values back into an object. Other than that it is not possible, because the initial created tensors are not.

**Comment by LeoBuron** [comment]
I can't you still have requested changes

File: tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py (hunk 9 above)
Comment (glencoe) [reply tests/unit_tests/nn/quantized_grads/fixed_point/quantize_to_fixed_point_test.py:79]: Oh... do you mean the owning module would have to reconstruct the config everytime?

Sorry, i should first approve this i guess.
I believe you that this is the way to do it, i just don't think i completely got why.
But i'll ask you again at another time, i'm curious 😉 .

**Comment by glencoe** [comment]
> I can't you still have requested changes

What i mean is, there are now three merge commits in this PR. If possible, please rebase this branch on main instead of merging main into it.
